### ISSUE TITLE
log errors only

### ISSUE THREAD

**Issue by pkpowell** [imapsync/imapsync#221]
Is it possible to log only errors?
Our LOG_imapsync dir grows about 500MB per pass (we have loads of public folder that get filtered from the sync but clog up the log)!

**Comment by gilleslamiral** [comment]
> Is it possible to log only errors?

Not directly with imapsync.
It's possible to keep only the last 300 lines of a command output or a file on Linux:

    imapsync ... --nolog | tail -300  > loglast300.txt 

or

    cat bigfile.txt  | tail -300  > loglast300.txt 

> Our LOG_imapsync dir grows about 500MB per pass (we have loads of public folder that get filtered from the sync but clog up the log)!

I don't get the point. 
When a folder is excluded, it shouldn't take many lines about it. 
How the clog lines look like?

**Comment by pkpowell** [comment]
Thanks, will try that.

Here's an excerpt from the log (header with command used below). Does this maybe relate to how Kerio deals with public folders?

Excluding folders matching pattern ^Public Folders.*
[Public Folders/000 Lehrlinge_2]
[Public Folders/000 Lehrlinge_2/aus]
[Public Folders/000 Lehrlinge_2/ein]
... (5096 entries)

This is followed by 5096 entries that look like this 

* LIST (\Noselect) "/" "Public Folders"
* LIST () "/" "Public Folders/000 Lehrlinge_2"
* LIST () "/" "Public Folders/000 Lehrlinge_2/aus"
* LIST () "/" "Public Folders/000 Lehrlinge_2/ein"
...

and then again like this
[Public Folders/000 Lehrlinge_2]
[Public Folders/000 Lehrlinge_2/aus]
[Public Folders/000 Lehrlinge_2/ein]
...

So it seems in total they're all listed 3 x.
Total file size in this case is 1.6MB, line count 15812

host1 is Kerio
host2 is Office365


Here is imapsync 1.977 on host mail.lan, a darwin system with 2.8/3.8 free GiB of RAM
with Perl 5.12.3 and Mail::IMAPClient 3.42
Transfer started at Thu Jun 11 08:52:01 2020
PID is 63803 my PPID is 63801
Log file is LOG_imapsync/2020_06_11_08_52_01_778_user1@boltshauser.info_user2@boltshauser.info.txt ( to change it, use --logfile path ; or use --nolog to turn off logging )
Load is 2.98 2.84 2.66 on 6 cores
Current directory is /Users/ard/migrator
Real user id is ard (uid 501)
Effective user id is ard (euid 501)
$RCSfile: imapsync,v $ $Revision: 1.977 $ $Date: 2019/12/23 20:18:02 $ 
Command line used, run by perl:
/usr/local/bin/imapsync --host1 mail.boltshauser.info --user1 user1@boltshauser.info --password1 MASKED --authmech1 X-MASTERAUTH --host2 outlook.office365.com --user2 user2@boltshauser.info --password2 MASKED --authuser2 philip.powell@boltshauser.info --exclude ^Public Folders.* --exclude ^Trash.* --exclude ^Junk.* --exclude ^Deleted Items.* --exclude ^(~).* --automap --addheader --noreleasecheck

**Comment by gilleslamiral** [comment]
Thanks for your input.
I don't intend to filter those outputs with an option. That would be ridiculous since a grep -v can do it easily. But I intend to add another very small log output that will summarize the current sync  during its execution. The last state of this file will be what you search. No deadline for it yet

**Comment by alfonsrv** [comment]
I agree. Logging is way too verbose. If the application has to be started multiple times it just creates a lot of unnecessary overhead.

A minimal output would be desirable (just showing transfers without any of the other stuff) and having the option to only pipe errors with `2> errors.txt`

**Comment by gilleslamiral** [comment]
> Logging is way too verbose. If the application has to be started multiple times it just creates a lot of unnecessary overhead.
> 

The problem with less logging is that I added each output line based on experience. And not from one-time experiences but from most frequent needs. So, what should I remove?

> A minimal output would be desirable (just showing transfers without any of the other stuff) 

What is the other stuff you want to avoid?
Besides, showing transfers is the vast size output on big mailboxes, not the rest.

> and have the option to only pipe errors with `2> errors.txt`

There is no STDERR output in imapsync and I won't use it, I actually removed it intentionally. 
The following

     imapsync  ...  --nolog | grep "^Err" | tee errors.txt

will give what you want.
Good luck with understanding most problems with this method, I wouldn't use it.

**Comment by alfonsrv** [comment]
> What is the other stuff you want to avoid?

Everything. I just want to see copied / deleted messages.

> So, what should I remove?

In my opinion the login logs + folder listings for both hosts as well as everything else associated with `imapsync` version, checking for "new releases" should be optional. So basically everything before actually copying the messages – to me – feels a bit too verbose. Between subsequent restarts that information is usually redundant. Alternatively a `--minimal` flag would toggle less verbose logs.

Folder mapping and listing should even be a dedicated flag entirely used in conjunction with `--dry`.

> `imapsync  ...  --nolog | grep "^Err" | tee errors.txt`

I tried something like that. However I want application errors – not SSL errors or other stuff `imapsync` does a phenomenal job auto-recovering at. Like `failed to login to host2`, `message XYZ could not be copied`, `could not delete message X`, `...`

**Comment by gilleslamiral** [comment]
> In my opinion the login logs + folder listings for both hosts as well as 

It is sometimes necessary to have them.

> everything else associated with imapsync version,

When someone reports me a log, I need the imapsync release because understanding what imapsync does is related to its release number.

> checking for "new releases" should be optional. 

It is. Use --noreleasecheck
https://imapsync.lamiral.info/#NUMBERS


> So basically everything before actually copying the messages – to me – feels a bit too verbose. 

I agree imapsync is verbose, it says what it does along its run. That's on purpose.

> Between subsequent restarts that information is usually redundant. 

We don't log the same for the "usually", we log, often the same, for the "unusually", the not so often.

> Alternatively a --minimal flag would toggle less verbose logs.

It's a long-standing request: 
https://imapsync.lamiral.info/TODO
```
SUGGESTED 2016_06_22 by Gilles
Add  --quiet mode
Make --quiet mode being switched by a signal
...
Add Rick Romero patch with
--quiet   No output at all
--showstats  intended for use with --quiet
```

No timeline...

> I want application errors – not SSL errors or other stuff imapsync does a phenomenal job auto-recovering at. Like failed to login to host2, message XYZ could not be copied, could not delete message X

I often need the context, which is the normal stuff.

**Comment by Micsi** [comment]
I am (once again) investigating this.
As I am using imapsync from time to time to transfer Mailservers with several mailboxes, I am using Skripts to do the transfer in batches. But the existing logging unfortunately is not that helpful in my experience. It is just too much to be helpful - at least for a human… therefore I think a --quiet option would be really helpful for all the cases where there is no problem at all. If a problem occurs, I would love to run the command with full output to see what's happening.

As it's not even one out of 10 runs with problems, I think this would really be a big help.

And you still have the Log-files, which you CAN disable. For me, it would be useful to have an interactive output quiet, just with error info, and the logfile available to check for details - or just run it again with full output…

Best regards, thanks for the wonderful tool,

Jochen

**Comment by gilleslamiral** [comment]
I'm too lazy to code a  `--quiet` option and think about what to quiet.
